src/fsharp/FSharp.Core/prim-types.fs

```diff
@@ -1505,12 +1636,179 @@ namespace Microsoft.FSharp.Core
                     override iec.Equals(x:obj,y:obj) = GenericEqualityObj true iec (x,y)  // ER Semantics
                     override iec.GetHashCode(x:obj) = raise (InvalidOperationException (SR.GetString(SR.notUsedForHashing))) }
 
+            let canUseDefaultEqualityComparer (ty:Type) =
+                let processed = System.Collections.Generic.HashSet ()
+
+                let rec recurse idx (types:array<Type>) =
+                    if idx = types.Length then true
+                    else
+                        let ty = get types idx
+                        if not (processed.Add ty) then
+                            recurse (idx+1) types
+                        else
+                            let isValidGenericType ifNotType fullname =
+                                if not (ty.IsGenericType && ty.GetGenericTypeDefinition().FullName.Equals fullname)
+                                then ifNotType
+                                else recurse 0 (ty.GetGenericArguments ())
+                    
+                            // avoid any types that need special handling in GenericEqualityObj
+                            true 
+                            && ty.IsSealed  // covers enum and value types
+                            && not ty.IsArray
+                            && not (ty.Equals typeof<String>)
+                            && not (ty.Equals typeof<Decimal>)
+                            && not (ty.Equals typeof<float>)
+                            && not (ty.Equals typeof<float32>)
+                            && isValidGenericType true "System.Nullable`1"
+                            && not (typeof<IStructuralEquatable>.IsAssignableFrom ty
```

Comment:
Not too sure what your saying, I can't use EqualityComparer.Default on records and union types because they have IStruturalEquality which has different ER/PER symantics:

    type Boo = Boo of float
    let nan = Boo System.Double.NaN
    printfn "%b" (HashIdentity.Structural.Equals (nan, nan))
    printfn "%b" (System.Collections.Generic.EqualityComparer.Default.Equals (nan, nan))

    // false
    // true

They are de-virtualized by the compiler under normal conditions - none of the things I'm changing should change generated code (which is borne out by no changed to the IL as part of the test suite)